ts/src/base/Exchange.ts

```diff
@@ -3609,26 +3610,59 @@ export default class Exchange {
         return this.handleOptionAndParams (params, methodName, 'marginMode', defaultValue);
     }
 
-    throwExactlyMatchedException (exact, string, message) {
-        if (string in exact) {
-            throw new exact[string] (message);
+    getExceptionMarketType (originUrl) {
+        if (originUrl !== undefined) {
```

Comment:
@ttodua I'm afraid this solution won't scale very well because sometimes these API divisions are just to organize the data better, check for instance bitget, gate, bingx, etx
![image](https://github.com/ccxt/ccxt/assets/43336371/814bb28c-4a70-4827-8e94-da85a0d5a65e)

What if instead, we keep `throwByUrl` and inside each class we define the method `getExceptionsByUrl`, and inside we each exchange we decide how to select the correct map, example:

```Typescript
    getExceptionsByUrl(url) {
        if (url.startsWith('eapi')) {
            return this.exceptions['option'];
        }  else if (url.startsWith('v1') || url.startsWith('public')) {
            return this.exceptions['spot'];
        }
        (...)
        return undefined;
    }
```

i think this is simpler and therefore a bit easier to understand and we remove the need of `getExceptionMarketType` and `exceptionTypeMappings`.

wdyt?

Reply:
for me, JSON-ification is the one that is easier imo, more structural and more unifiish, whereas I think that hardcoded things are on opposite side in those aspects. my approach just adds `exceptionTypeMappings`, so there will be no need to manually add `getExceptionsByUrl` method to 100 exchanges (with their method body manually hardcoded in complex if/else tree, and accessing the dictionaries there with manual string-keys).  but it's just me, as you tell me, I'll do so.

Reply:
specifically about bitget (or similar excahnges like on your screenshot) I think that's the problem in the object, probably instead of:
```
spot: http://api.{hostname}
...
```
we needed to have:
```
spot: https://api.{hostname}/spot
mix: https://api.{hostname}/mixapi_whatever
...
```
but, in given situation, i think it would need to be fixed those schemes on those exclusion exchanges, and then could be pushed the exception-per-market PRs.

Reply:
however, again, this is my view/thoughts to "oppose", but please you make a decision which you still see the better approach and let me know in next comment, so I'll do that way right away.

Reply:
@ttodua 

> my approach just adds exceptionTypeMappings, so there will be no need to manually add getExceptionsByUrl


what's the difference between adding one or the other? Ofc you might argue that `exceptionTypeMappings` is more generic but `getExceptionsByUrl` is a pretty simple method as well and it's more flexible in case the exchange has some weird specificity 


> i think it would need to be fixed those schemes on those exclusion exchanges, and then could be pushed the exception-per-market PRs.

That's not desirable, I mean changing the structure of the API of a lot of exchanges to introduce this feature when we have an alternative seems too risky and labourious

Reply:
I've seen now gate. it seems a bit harder to wrap into this approach directly, however, for gate we could have in-exchange solution like you said (and don't use the base methods & mappings). 
idk as you say

Reply:
oky, doing that way ✌️

Reply:
sorry missed to answer. imo, about the differences
- there were several differences.  first, it was straightforward json map, without need to go through if/else logic, and actually, and it didn't need any logic/coding to have inside derived classes, as . second, that was more structural, exposing ability to use it in reverse (who know, for further customization if users ever want).
- about custom/complex exchanges: neither it was a problem, because for those complex exchanges we could have used "by method" approach. moreover, I think eventually `bitget` like exchanges would still need to have their api-tree fixed, because currently it's basically listed out in `sign`, whereas it should have definitely been in `api` tree, where anyone can see it directly which directory refers to which url.

but sure, nothing critical, pushed the updated version, plz take a look into it.